native/c/zjson.hpp

```diff
@@ -397,20 +534,193 @@ class ZJson
+    {
+      throw std::runtime_error("No deserializer registered for type");
+    }
+
+    // Get the specified path
+    JsonValueProxy path_proxy = (*const_cast<ZJson *>(this))[path];
+
+    deserializer(result, path_proxy);
+    return result;
+  }
+
+  // Helper method to check if a type can be serialized
+  template <typename T>
+  static constexpr bool canSerialize()
+  {
+    return is_serializable<T>::value && SerializationRegistry<T>::hasDeserializer();
+  }
 };
 
+// Macro for easy struct registration (after ZJson class definition)
+#define ZJSON_SERIALIZABLE(StructType, ...)                                                                        \
+  template <>                                                                                                      \
+  struct is_serializable<StructType> : std::true_type                                                              \
+  {                                                                                                                \
+  };                                                                                                               \
+  namespace                                                                                                        \
+  {                                                                                                                \
+  struct StructType##_Registrar                                                                                    \
+  {                                                                                                                \
+    StructType##_Registrar()                                                                                       \
+    {                                                                                                              \
+      SerializationRegistry<StructType>::registerDeserializer(                                                     \
+          [](StructType &obj, const ZJson::JsonValueProxy &json) { deserialize_fields(obj, json, __VA_ARGS__); }); \
+    }                                                                                                              \
+  };                                                                                                               \
+  static StructType##_Registrar StructType##_registrar_instance;                                                   \
+  }
+
+// Serialization registry implementation (after ZJson class definition)
+template <typename T>
+class SerializationRegistry
+{
+public:
+  using DeserializerFunc = std::function<void(T &, const ZJson::JsonValueProxy &)>;
+
+  static void registerDeserializer(DeserializerFunc func)
+  {
+    getDeserializer() = func;
+  }
+
+  static DeserializerFunc &getDeserializer()
+  {
+    static DeserializerFunc deserializer;
+    return deserializer;
+  }
+
+  static bool hasDeserializer()
+  {
+    return static_cast<bool>(getDeserializer());
+  }
+};
```

Comment:
All the functions here are static - what are your thoughts on making this a Singleton?


```suggestion
#include "singleton.hpp"
template <typename T>
class SerializationRegistry : public Singleton<SerializationRegistry<T>>
{
  friend class Singleton<SerializationRegistry<T>>;
public:
  using DeserializerFunc = std::function<void(T &, const ZJson::JsonValueProxy &)>;

  static void registerDeserializer(DeserializerFunc func)
  {
    getDeserializer() = func;
  }

  static DeserializerFunc &getDeserializer()
  {
    static DeserializerFunc deserializer;
    return deserializer;
  }

  static bool hasDeserializer()
  {
    return static_cast<bool>(getDeserializer());
  }
};
```